[IDE] Report ambiguous cursor info results

ahoppen opened the pull request:
While we switched cursor info to solver-based I actually missed to return results in ambiguous cases. This PR fixes that oversight.

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen committed:
[CursorInfo] Fix issues found by the stress tester

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
All stress tester failures are one of the following:
- We fail to do local rename because the solver-based CursorInfo is able to resolve the cursor where the AST-based cursor info that we need for refactorings can’t. This is an improvement because previously neither rename nor cursor info worked, now only cursor info is working
- There is a crash in SymbolGraph when doing Cursor Info on `index` in the stdlib. I think we can address this in a follow-up PR so we can merge this PR. A reduced test case is
```swift
func foo(hex: String) {
  // RUN: %sourcekitd-test -req=cursor -req-opts=retrieve_symbol_graph=1 -pos=%(line + 1):7 %s -- %s
  hex.index
}
```

rintaro commented on lib/IDE/CursorInfo.cpp:
```diff
@@ -273,13 +276,33 @@ class CursorInfoTypeCheckSolutionCallback : public TypeCheckCompletionCallback {
+  /// and thus the constraint system solution doesn't know about the
+  /// \c UnresolvedDeclRefExpr. Instead, we find the expression to resolve in
+  /// the source file again after expression pre-check has run.
+  SourceFile &SrcFile;
```
Can this be the decl context instead? That should be faster.

ahoppen replied:
Good idea.

ahoppen commented on tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp:
```diff
@@ -1891,12 +1921,18 @@ static void deliverCursorInfoResults(
-                            Invoc, Diagnostic, /*PreviousSnaps=*/{},
-                            Results->DidReuseAST, Receiver);
+    CursorInfoData Data;
+    for (auto ResolvedCursorInfo : Results.getResult().ResolvedCursorInfos) {
```
Oh, I completely forgot that I added `operator->`

ahoppen commented:
@swift-ci Please smoke test

xedin commented on lib/Sema/TypeCheckConstraints.cpp:
```diff
@@ -416,7 +416,11 @@ TypeChecker::typeCheckTarget(SolutionApplicationTarget &target,
 
-  if (DiagnosticSuppression::isEnabled(Context.Diags))
+  /// If we have a SolutionCallback, we are inspecting constraint system
+  /// solutions directly and thus also want to receive ambiguous solutions.
```
This is actually interesting because there could be a number of _valid_ solutions produced by the solver and I'm guessing we don't really want to run `salvage` in that case, right? It feels like this check needs to be embedded into the solver itself somehow...

ahoppen replied:
You mean that we should just allow fixes straight away like we do for code completion? 

I deliberately wanted to keep this as close to the normal type checking as possible (without all the code-completion specific rules of weaker filtering etc.) because if normal type checking reaches a single, definite result, we also want to report that as the single result for cursor info (for now). Also, in contrast to code completion, there is a reasonable chance that the code is in a valid state and thus it makes sense to run normal type-checking first and only salvage if we didn’t find a solution or the solutions were ambiguous.

ahoppen replied:
I added this check to `ConstraintSystem::solve` now. Could you take another look at it?

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen committed:
[IDE] Support reporting multiple cursor results

This hooks up the cursor info infrastructure to be able to pass through multiple, ambiguous results. There are still minor issues that cause solver-based cursor info to not actually report the ambiguous results but those will be fixed in a follow-up PR.

ahoppen committed:
[IDE] Report ambiguous cursor info results

ahoppen commented:
@swift-ci Please smoke test

rintaro commented on tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp:
```diff
@@ -1170,65 +1169,95 @@ static bool passCursorInfoForDecl(
+    return false;
+  }
+
+  Data.AvailableActions.insert(Data.AvailableActions.end(),
+                               KnownRefactoringInfo.begin(),
+                               KnownRefactoringInfo.end());
```
This could be `append(KnownRefactoringInfo.begin(), KnownRefactoringInfo.end())`, but I see the existing code used `insert`.

ahoppen commented:
@swift-ci Please smoke test
